### PR TITLE
ap-4509: add linked_applications table, model and associations

### PR DESCRIPTION
**Pull request by kmahern**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4381)

Added migration to add linked_applications table.
Added LinkedApplication model.
Added specs and LinkedApplications factory

TODO:

implement validations on LinkedApplications, however, this is to be done as part of ap-4510.

Describe what you did and why.

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: spec/models/legal_aid_application_spec.rb [hunk 1]
```diff
@@ -1745,6 +1745,65 @@
     end
   end
 
+  describe "#associated_applications" do
```
Comment (jsugarman) [inline spec/models/legal_aid_application_spec.rb:1748]: Please see comment and suggestion for #lead_application spec.

File: spec/models/legal_aid_application_spec.rb [hunk 2]
```diff
@@ -1745,6 +1745,65 @@
...
+    end
+
+    it "calling destroy! deletes the linked_application record but not the linked appication itself" do
+      linked_application = LinkedApplication.create!(lead_application:, associated_application:, link_type_description: "link", link_type_code: "LNK")
+      lead_application.destroy!
+      expect { described_class.find(lead_application.id) }.to raise_error ActiveRecord::RecordNotFound
+      expect { LinkedApplication.find(linked_application.id) }.to raise_error ActiveRecord::RecordNotFound
+      expect(described_class.find(associated_application.id)).to eq associated_application
+    end
+  end
```
Comment (jsugarman) [inline spec/models/legal_aid_application_spec.rb:1805]:
```ruby
    context "when destroying a lead application" do
      before do
        create(:linked_application, :family, lead_application: application_one, associated_application: application_two)
      end

      it "destroys the link but not the associated application" do
        expect { application_one.destroy! }
          .to change { application_two.reload.lead_application }
              .from(application_one)
              .to(nil)
      end
    end
```

File: spec/models/legal_aid_application_spec.rb [hunk 3]
```diff
@@ -1745,6 +1745,65 @@
...
+  describe "#lead_application" do
+    let(:lead_application) { create(:legal_aid_application) }
+    let(:associated_application) { create(:legal_aid_application) }
+
+    it "returns the lead_application" do
+      linked_application = LinkedApplication.create!(lead_application:, associated_application:, link_type_description: "link", link_type_code: "LNK")
+
+      expect(associated_application.lead_application).to eq lead_application
+      expect(associated_application.lead_linked_application).to eq linked_application
+    end
```
[the remark this answers is not in the document]
Comment (jsugarman) [reply spec/models/legal_aid_application_spec.rb:1792]: factory trait added too
```ruby
    trait :family do
      link_type_code { "FAMILY" }
      # link_type_description { "Family" }
    end
```

File: db/migrate/20231009145448_create_linked_applications.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+class CreateLinkedApplications < ActiveRecord::Migration[7.0]
+  def change
+    create_table :linked_applications, id: :uuid do |t|
+      t.references :lead_application, type: :uuid, null: false
+      t.references :associated_application, type: :uuid, null: false  
+      t.string :link_type_description, null: false
```
[the remark this answers is not in the document]
Comment (kmahern) [reply db/migrate/20231009145448_create_linked_applications.rb]: 👍 agreed removed.

File: spec/models/legal_aid_application_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmahern) [reply spec/models/legal_aid_application_spec.rb]: 👍 have amended spec as suggested

File: spec/models/legal_aid_application_spec.rb (hunk 2 above)
Comment (kmahern) [reply spec/models/legal_aid_application_spec.rb]: 👍 have amended as suggested

File: app/models/linked_application.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class LinkedApplication < ApplicationRecord
+  belongs_to :lead_application, class_name: "LegalAidApplication"
+  belongs_to :associated_application, class_name: "LegalAidApplication"
+
+  validate :cannot_link_self
+
+private
+
+  def cannot_link_self
+    errors.add(:linked_application_ref, "You cannot link an application to itself.") if lead_application == associated_application
```
[the remark this answers is not in the document]
Comment (kmahern) [reply app/models/linked_application.rb]: 👍 have moved to translation

File: spec/models/legal_aid_application_spec.rb [hunk 6]
```diff
@@ -1745,6 +1745,65 @@
...
+
+    it "does not allow an a application to be linked to a nonexistent application" do
+      expect { LinkedApplication.create!(lead_application:, associated_application_id: "11111111-2222-3333-4444-555555555555", link_type_description: "link", link_type_code: "LNK") }.to raise_error ActiveRecord::RecordInvalid, "Validation failed: Associated application must exist"
+    end
```
[the remark this answers is not in the document]
Comment (kmahern) [reply spec/models/legal_aid_application_spec.rb]: 👍 have removed

File: spec/models/legal_aid_application_spec.rb (hunk 1 above)
Comment (kmahern) [reply spec/models/legal_aid_application_spec.rb:1748]: 👍 have amended as per suggestion

File: app/models/linked_application.rb [hunk 7]
```diff
@@ -7,6 +7,6 @@ class LinkedApplication < ApplicationRecord
 private
 
   def cannot_link_self
-    errors.add(:linked_application_ref, "You cannot link an application to itself.") if lead_application == associated_application
+    errors.add(:linked_application_ref, I18n.t("activemodel.errors.models.linked_application.attributes.application_cannot_be_linked_to_itself.message")) if lead_application == associated_application
```
[the remark this answers is not in the document]
Comment (jsugarman) [reply app/models/linked_application.rb]: done

**Pull request merged**